### PR TITLE
Long mode

### PR DESCRIPTION
**Pull request by RoyShulman**

### CODE REVIEW

File: debug.sh [hunk 1]
```diff
@@ -1,6 +1,6 @@
 rm ./builder/target/bootloader.img
 rm -r target/x86_64-bootloader
-cargo xbuild --target ./x86_64-bootloader.json
-cd ./builder && cargo run ../target/x86_64-bootloader/debug/robustloader ../kernel.elf # TODO: remove this example non existent kernel.elf
+cargo xbuild --target ./x86_64-bootloader.json --release
+cd ./builder && cargo run ../target/x86_64-bootloader/release/robustloader ../kernel.elf # TODO: remove this example non existent kernel.elf
 cd ../ && qemu-system-x86_64 -s ./builder/target/bootloader.img &
-gdb -ex "target remote localhost:1234" -ex "symbol-file ../target/x86_64-bootloader/debug/robustloader"
+gdb -ex "target remote localhost:1234" -ex "symbol-file ../target/x86_64-bootloader/release/robustloader"
```
Comment (talbor49) [inline debug.sh:6]: Why is debug.sh compiling with release?

File: src/asm/stages/stage_2.asm [hunk 2]
```diff
@@ -10,19 +10,31 @@ stage_2:
...
-    lea eax, [stage_3]
-    push eax
-    retf
-    h:
-    	jmp h
+    lea si, return_32_bit_mode_str
+    call print_string_16
 
+    call enter_protected_mode 	# This function only sets the data segmentss
+    push 0x8					# Push the code segment in the gdt
```
Comment (talbor49) [inline src/asm/stages/stage_2.asm:33]: maybe set 0x8 to a const? this magic is unclear even with the comment

File: src/asm/stages/stage_3.asm [hunk 3]
```diff
@@ -0,0 +1,199 @@
+.section .boot, "awx"
+.intel_syntax noprefix
+.code32
+
+
+# This stage checks for 64 bit, Sets up paging and enters long mode
+
+stage_3:
+	mov ax, 0x10
```
Comment (talbor49) [inline src/asm/stages/stage_3.asm]: Please explain this magic with comments, or even better with a descriptive name for the const if possible

File: src/asm/stages/stage_3.asm [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+	# first zero out the page table!
+	lea edi, [__page_table_start]
+	lea ecx, [__page_table_end]
+	sub ecx, edi
+	shr ecx, 2 # one stosd zeros 4 bytes
+	xor eax, eax
+	rep stosd # zero out entire page table
+
+	push ecx
```
Comment (talbor49) [inline src/asm/stages/stage_3.asm]: Why are we saving ecx here? It was just zeroed

File: src/asm/stages/stage_3.asm [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	pop ecx
+	ret
+
+
+enable_paging:
+	# This function enables paging by writing the address of _pmlt4 to cr3
+	# setting the pae flag in cr4
+	# setting the long mode bit in EFER
+	# and setting the paging bit in cr0
+    wbinvd # write back cache
```
Comment (talbor49) [inline src/asm/stages/stage_3.asm:131]: Fix indentation here and in the rest of the code

File: src/asm/stages/stage_3.asm [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	or eax, 1 << 5 	# Enable the PAE flag in cr4 (The physical address extesion)
+	mov cr4, eax
+
+	mov ecx, 0xc0000080 # Set the long mode bit in EFER MSR (Extended Feature Enable Register)
+	rdmsr 				# Read model specific register specified in ecx into edx:eax
+	or eax, 1 << 8		# Enable long mode bit
+	wrmsr 				# Write into model specific register
+
+	mov eax, cr0	 	# Enable paging 
+	or eax, 1 << 31
```
Comment (talbor49) [inline src/asm/stages/stage_3.asm:147]: Write which bit we are turning on here

File: src/asm/stages/stage_3.asm [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+.include "src/asm/utils/32_print_string.asm"
+
+cpuid_not_supported_str:		.asciz	"cpuid is not supported, no way to check long mode"
+long_mode_not_supported_str:	.asciz	"long mode is not supported!"
+paging_enabled_str:				.asciz  "Paging is officaly enabled!"
+
+.code64
+long_mode_start:
+	mov rax, 0x2f592f412f4b2f4f
```
Comment (talbor49) [inline src/asm/stages/stage_3.asm:195]: what the fuck? comment this (printing OKAY or whatever)

File: src/asm/stages/stage_3.asm [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+	mov ecx, 0xc0000080 # Set the long mode bit in EFER MSR (Extended Feature Enable Register)
+	rdmsr 				# Read model specific register specified in ecx into edx:eax
+	or eax, 1 << 8		# Enable long mode bit
+	wrmsr 				# Write into model specific register
+
+	mov eax, cr0	 	# Enable paging 
+	or eax, 1 << 31
+	mov cr0, eax
+
+	pop ecx
```
Comment (talbor49) [inline src/asm/stages/stage_3.asm]: Why are we popping ecx here?

File: src/asm/stages/stage_3.asm [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+	lidt [zero_idt]
+
+	call setup_pae_paging
+	call enable_paging
+
+	lea si, paging_enabled_str
+	call print_string_32
+
+	lgdt gdt_64_pointer
+	push 8
```
Comment (talbor49) [inline src/asm/stages/stage_3.asm]: Before you pushed 0x8, try to be consistent with the counting bases so it's clear that it is for the same purpose

File: src/asm/utils/32_print_string.asm [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    pusha 
+    mov edx, 0xb8000
+print_string_32_loop:
+    mov al, [esi]               # Store the char at ebx in AL
+    mov ah, 0x4f                # Store the attributes in AH
+
+    cmp al, 0                   # Check if we hit the null terminator
+    je print_string_32_done
+
+    mov [edx], ax               # Store the char and the attributes at the correct index in the VGA
```
Comment (talbor49) [inline src/asm/utils/32_print_string.asm:20]: this is cool!

File: src/main.rs [hunk 11]
```diff
@@ -20,14 +27,32 @@ fn panic(_info: &PanicInfo) -> ! {
...
 #[no_mangle] // don't mangle the name of this function
-pub extern "C" fn main_loop() -> ! {
+pub unsafe extern "C" fn stage_4() -> ! {
     // this function is the entry point, since the linker looks for a function
     // named `_start` by default
+    // set stack segment
+    asm!("mov bx, 0x00
+          mov ss, bx" ::: "bx" : "intel");
+    printer::Printer.clear_screen();
+    write!(printer::Printer, "RUNNING FROM RUST YEAH  :D :))))))) !!!!@#!@#!@#!@#!@#!@#").unwrap();
```
Comment (talbor49) [inline src/main.rs:56]: Oh yeah!!!!!!!!! :)))

File: src/printer/vga_text_80x25.rs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    }
+}
+
+impl Write for Printer {
+    fn write_str(&mut self, s: &str) -> Result {
+        let vga_buffer = Self::vga_buffer();
+        for byte in s.bytes() {
+            let index = CURRENT_OFFSET.fetch_add(2, Ordering::Relaxed);
+            vga_buffer[index] = byte;
+            vga_buffer[index + 1] = 0x4f;
```
Comment (talbor49) [inline src/printer/vga_text_80x25.rs]: Documentate this magic

File: src/asm/stages/stage_2.asm (hunk 2 above)
Comment (RoyShulman) [reply src/asm/stages/stage_2.asm:33]: no idea how to make consts in this assembly...

File: debug.sh (hunk 1 above)
Comment (RoyShulman) [reply debug.sh:6]: runs much faster with release, with no real difference

File: src/asm/stages/stage_3.asm (hunk 3 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm]: fixed with comments

File: src/asm/stages/stage_3.asm (hunk 5 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm:131]: looks fine to me...

File: src/asm/stages/stage_3.asm (hunk 6 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm:147]: fixed

File: src/asm/stages/stage_3.asm (hunk 7 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm:195]: fixed

File: src/asm/stages/stage_3.asm (hunk 8 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm]: to save the state of ecx before we called the function

File: src/asm/stages/stage_3.asm (hunk 9 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm]: fixed

File: src/asm/utils/32_print_string.asm (hunk 10 above)
Comment (RoyShulman) [reply src/asm/utils/32_print_string.asm:20]: yes!

File: src/printer/vga_text_80x25.rs (hunk 12 above)
Comment (RoyShulman) [reply src/printer/vga_text_80x25.rs]: this is the same as the print_string_32 function just in rust

File: src/asm/stages/stage_3.asm (hunk 4 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm]: fixed

File: src/asm/stages/stage_2.asm (hunk 2 above)
Comment (talbor49) [reply src/asm/stages/stage_2.asm:33]: actually i have no clue either, "GDT_CODE_SEGMANT equ 8" didn't work :(

File: src/asm/stages/stage_3.asm (hunk 5 above)
Comment (talbor49) [reply src/asm/stages/stage_3.asm:131]: You're right, it looks weird on github but it's fine in the code, sry :)

File: src/asm/stages/stage_3.asm (hunk 8 above)
Comment (talbor49) [reply src/asm/stages/stage_3.asm]: yeah but why do you want to save ecx specifically? I couldn't find a use of it later

File: src/printer/vga_text_80x25.rs (hunk 12 above)
Comment (RoyShulman) [reply src/printer/vga_text_80x25.rs]: fixed consts

File: src/asm/stages/stage_3.asm (hunk 8 above)
Comment (RoyShulman) [reply src/asm/stages/stage_3.asm]: you are right. deleted it

**Pull request merged**